src/facade/reply_builder.cc

```diff
@@ -856,11 +856,13 @@ void ReqSerializer::SendCommand(std::string_view str) {
 
 void RedisReplyBuilder2Base::SendNull() {
   ReplyScope scope(this);
+  has_replied_ = true;
```

Comment:
please set this flag when we drop a scope, not manually everywhere

Reply:
I am aware but this would break `MCReplyBuilder2`. `ReplyScope` accepts a `SinkReplyBuilderv2`. However, `has_replied_` belongs to a different class, `SinkReplyBuilder`. The problem is that `this` argument for `RedisReplyBuilder2` would work just fine because the `RedisReplyBuilder2Base` inherits from: ` SinkReplyBuilder2,  RedisReplyBuilder` so it's safe to cast it to `RedisReplyBuilder`. However this is not the case for `MCReplyBuilder2`. 

I agree `has_replied_` should be part of `ReplyScope`. For this we need:
1. Clean up the hierarchy of inheritence
2. Or even introduce a new has_replied_ flag. However this would no work very well because we use `SinkReplyBuilder*` which would poll the base class flag and we also can't really remove it because we don't want to break previously existed classes

Reply:
it should've been just part of the SinkReplyBuilder2

Reply:
uh I was not expecting you would reply early :( I will push a quick change on that once I get a 5 minute free time :)

Although that means we break `SinkReplyBuilder`. Imo what we can do is just make `HasReplied` virtual and it will solve all of our problems. I will take care of this.

P.s. glad to see you commenting :heart: